### ISSUE TITLE
Bug - Get-LongDirectorySize - You cannot call a method on a null-valued expression

### ISSUE THREAD

**Issue by Zuldan** [v2kiran/PSAlphaFS#6]
PS C:\> Get-LongDirectorySize -Path 'C:\Temp\'-Recurse -ContinueonError

```
You cannot call a method on a null-valued expression.
At C:\Users\gshipp\OneDrive\Powershell\Profile\Modules\PSAlphaFS\PSAlphaFS.psm1:1360 char:6
+         if(-not $DirObject::Exists($Path))
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
```

PS C:\Get-Module -ListAvailable -Name PSAlphaFS
```

    Directory: C:\Users\gshipp\OneDrive\Powershell\Profile\Modules

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     2.0.0.0    PSAlphaFS                           {Copy-LongItem, DisMount-LongShare, Get-LongChildItem, Get-LongDirectorySize...}
```

PS C:\$PSVersionTable

```
Name                           Value
----                           -----
PSVersion                      5.1.14409.1005
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14409.1005
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

**Comment by Zuldan** [comment]
Commenting out the following code makes the function work perfectly.

```
		if(-not $DirObject::Exists($Path))
		{
			Write-Warning -Message ("Get-LongDirectorySize:`tPath '{0}' either is not a directory or it dosent exist." -f $Path)
			return
		}
```

PS C:\> Import-Module PSAlphaFS -Force
PS C:\> Get-LongDirectorySize -Path 'C:\Temp' -Recurse -ContinueonError

```
Key                      Value
---                      -----
Archive                 639399
Compressed                   0
Device                       0
Directory                83048
Encrypted                    0
Hidden                   16176
IntegrityStream              0
Normal                    3366
NoScrubData                  0
NotContentIndexed        11462
Offline                      0
ReadOnly                 17340
ReparsePoint                 0
SparseFile                   0
System                   15238
Temporary                  215
File                    642882
Total                   725930
Size              558002581657
SizeinGB                   520
```

**Comment by v2kiran** [comment]
@Zuldan - just tried this on 2 different systems and I am unable to reproduce this error. Are you using `-ContinueonError` because of subfolders in c:\temp that you dont have access to?  Also could you please paste the output from ($error[0] | fl * -force)

**Comment by Zuldan** [comment]
@v2kiran i include -ContinueonError because of habit. The folder actually only has a couple of files which I have full access to. The error occurs even if I do not use -ContinueonError.

I'm using Windows 8.1 x64 with .NET 4.6.2 and PS 5.1.14409.1005. I'm not sure if you have that exact configuration in a lab but I'm able to replicate the issue on a couple of other machines.

**Comment by Zuldan** [comment]
I also get the error when using Get-LongChildItem. Same line of code bombs out. This failed on 9x Windows 2012 R2 servers we have. They have .NET 4.6.2 and PS 5.1.14409.1005.

$DirObject::Exists is causing the problem.

PS C:\> Get-LongDirectorySize -Path D:\Home -Recurse
```
You cannot call a method on a null-valued expression.
At C:\Program Files\WindowsPowerShell\Modules\PSAlphaFS\PSAlphaFS.psm1:1360 char:6
+         if(-not $DirObject::Exists($Path))
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull

You cannot call a method on a null-valued expression.
At C:\Program Files\WindowsPowerShell\Modules\PSAlphaFS\PSAlphaFS.psm1:1376 char:3
+         $ResultHash = $DirObject::GetProperties( $Path, $dirEnumOptio ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
```

**Comment by Zuldan** [comment]
After commenting out the code around line 1360. Line 92 then breaks.

PS C:\Users\lab> Get-LongChildItem C:\Temp -Recurse
```
You cannot call a method on a null-valued expression.
At C:\Program Files\WindowsPowerShell\Modules\PSAlphaFS\2.0.0.1\PSAlphaFS.psm1:92 char:7
+             if(-not $PathFSObject::IsPathRooted($Path, $true))
+                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull

You cannot call a method on a null-valued expression.
At C:\Program Files\WindowsPowerShell\Modules\PSAlphaFS\2.0.0.1\PSAlphaFS.psm1:98 char:7
+ ...          if(-not ($FileObject::Exists($pItem) -or $DirObject::Exists( ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull

New-Object : Cannot find an overload for "FileInfo" and the argument count: "2".
At C:\Program Files\WindowsPowerShell\Modules\PSAlphaFS\2.0.0.1\PSAlphaFS.psm1:104 char:18
+ ... athObject = New-Object -TypeName Alphaleonis.Win32.Filesystem.FileInf ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [New-Object], MethodException
    + FullyQualifiedErrorId : ConstructorInvokedThrowException,Microsoft.PowerShell.Commands.NewObjectCommand
```
I put a break on line 92 

```
# Check if path is relative
if(-not $PathFSObject::IsPathRooted($Path, $true))
{
	$pItem = $PathFSObject::Combine($PWD, $pItem.TrimStart('.\'))
}
```
and $PathFSObject has nothing assigned to it.

```
[DBG]: PS C:\Program Files\WindowsPowerShell\Modules\PSAlphaFS\2.0.0.1>> $null -eq $PathFSObject
True
```

Something very odd is going on here. Maybe it's a scoping issue?

**Comment by v2kiran** [comment]
@Zuldan  - "failed on 9x machines" ..wow something is strange here...I have been insanely busy lately but I promise to look into this soon.
When I was writing this module I tested it on windows 10(my desktop) and some 2012 machines on azure.

**Comment by AndiBellstedt** [comment]
Had the same issue on using "Get-LongItem", "Get-LongChildItem" and "Remove-LongItem".
The strange thing on in my environment. The cmdlets are working fine on Powershell_ISE.exe but fail in powershell.exe.

I dig into the module manifest file of PSAlpha and find:
   ScriptsToProcess = @('helpers\PSAlphaFSHelpers.ps1')
In this script there is a scriptpart:
#Define Alphafs Class Shortcuts 

I assume there is a scoping problem with the variables. The variables in the helper file are not defined as global: or script: (not sure which is the better/right on !?)


**My workarround fixing the issue:**
Putting the helper variables in my own script before calling the Get-LongItem functions.
_$global:DirObject = [Alphaleonis.Win32.Filesystem.Directory]
$global:FileObject = [Alphaleonis.Win32.Filesystem.File]
........_

**My Environment (Win Srv 2012 R2):**
$PSVersionTable :
PSVersion                      4.0
WSManStackVersion              3.0
SerializationVersion           1.1.0.1
CLRVersion                     4.0.30319.42000
BuildVersion                   6.3.9600.17400
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0}
PSRemotingProtocolVersion      2.2

but same issue on:
PSVersion 5.0.10586.117
CLRVersion 4.0.30319.42000

**Comment by v2kiran** [comment]
@AndiBellstedt  - thank you for confirming the issue. I am able to replicate the behavior on a vanilla win 2012 r2 machine... you have the right idea I would go with "Global" because then the variable(s) would be present in the powershell instance instead of just the script....and now to fix it :)

**Comment by AndiBellstedt** [comment]
@v2kiran I'm not pretty sure if the scoping will work, but try to use "script" instead of "global". A friend told me some weeks before that this would be the better way, because variables with "script" scope would be only available in the module and are not fully presented to the console out of the module scope.

But all over - I am happy to give some help. :-)

**Comment by v2kiran** [comment]
ah I see...will do some tests with both scopes.

**Comment by Zuldan** [comment]
@v2kiran fantastic. Happy to help with testing.

**Comment by v2kiran** [comment]
Did some testing and found that when you use 'scriptstoprocess' in the module manifest the variables and functions  seem to be exist in the global scope. 
It dosent matter whether you use export-modulemember or 'functionstoexport' the private functions remain available.

After a few more tests I found that that the best way to handle private functions is to dot source them in the main psm1 file. this took care of both the variable and the function scope, the private functions & variables now dont leak into the session.

I pushed these changes to a new branch named "issue_6". if you guys @Zuldan  & @AndiBellstedt  find some time could you test it and see if it works for you?

**Comment by Zuldan** [comment]
@v2kiran I can confirm everything works perfectly with that build. Good work!

**Comment by v2kiran** [comment]
@Zuldan  - thanks for the confirmation. appreciate it!

**Comment by v2kiran** [comment]
fixed in v2.0.0.1
